code/contracts/ONEWallet2.sol

```diff
@@ -0,0 +1,176 @@
+
+    enum OperType { TRANSFER, RECOVERY, SET_LAST_RESORT_ADDR }
+
+    struct Commit {
+        bytes32 hash;          // h(eOTP || operType || ...params)  
+        uint timestamp;      // OTP must be valid in this time; not in the time of reveal 
+    }    
+    mapping(uint32 => Commit[]) public commits; // idx of OTP valid during a commitment => commitment data
+
+    event UnknownTransferError(address dest);
+    event UnknownOperationType(OperType otype);
+
+    constructor(bytes32 root_, bytes32 recoveryHash_, uint8 height_, uint8 interval_, uint32 t0_, uint32 lifespan_, address payable lastResortAddress_, uint256 dailyLimit_){                
+        root = root_; // Merkle root for OTPs
+        recoveryHash = recoveryHash_;
+        height = height_;
+        interval = interval_;        
+        t0 = t0_; 
+        lifespan_ = lifespan_;
+        lastResortAddress = lastResortAddress_;
+        dailyLimit = dailyLimit_; 
+        // lastRevealedOTPIdx = 0; // implicit
+        // lastExecutedTransfer = 0; // implicit
+        // lastCleanUpOfCommitsIdx = 0; // implicit
+    }
+
+    receive() external payable {}
+
+    /**
+     * This is a general commit and should work with all kinds of operations.
+     * Note that there can be multiple commits submitted in the current interval, while only the first matching one will be executed in reveal().
+     */
+    function commit(bytes32 hash) external {
+        uint32 submittedOTPIdx = uint32(block.timestamp) / interval - t0;
+        require(submittedOTPIdx > lastRevealedOTPIdx, "The OTP of the current interval was already revealed. It is not possible to submit commits in the current OTP interval anymore. Wait for the next interval.");
+        Commit memory nc = Commit(hash, block.timestamp);
+        commits[submittedOTPIdx].push(nc);                
+    }
+
+    /**
+     * This is a general reveal for all types of operation. Arguments contain all possible parameters as the union (only the relevant args for a particular operation are used).
+     * It cannot revert with not yet invalidated OTP, since the attacker can steal it + its  proof and misuse it for the malicious commit() and reveal().
+     */
+    function reveal(bytes32[] calldata neighbors, uint32 otpIdx, bytes32 eotp, OperType operType, address payable addr, uint amount) external returns (bool) {                        
+
+        // 1) Check time constrains of reveal
+        uint32 currentIdx = uint32(block.timestamp) / interval - t0; // check whether idx of OTP is timely
+        require(currentIdx - otpIdx <= MAX_REVEAL_DELAY, "Reveal is out of max allowed delay."); // here is OK to revert, coz if revert fails, submitted eotp + proof are invalidated already and cannot be misused anymore.
+
+        // 2) scan all the commits made within otpIdx interval and execute the first matching one        
```

Comment:
Assume a user made a commit and revealed within the same otpIdx interval (30 seconds in general, which is necessary for good user experience). After observing the revealed EOTP but before it is confirmed and executed by the blockchain nodes, an attacker could make another commit with the correct EOTP using different parameters of `operType`, `addr`, and `amount` and produce a valid hash in that commit. The attacker can then reveal its own commit at a higher gas cost before the user's reveal is processed. The attacker's reveal will succeed because it meets the requirement here: under the attacker's parameters, it will be the `first matching commit within the otpIdx interval`.

Reply:
That is true. It was very likely the reason why I put the parameters of an operation into the commit() in the original version instead of reveal(). The first matching one will be always the user's operation. Unfortunately, I did not make any comment to the code about it. So, with parameters in the commit(), it is secure but reveals user intention about the operation, which I still think does cause any problems in the scope of the standard wallet use cases.

Reply:
Then the current implementation (#v0.4.1, also current master branch) would be similar, except keys of `commits` are encoded as `commitHash` to hide time information and to prevent attacker from committing at future slots, and parameters are also encoded as `paramsHash` to hide user's intent, and to make the structure work for all types of operations in addition to `transfer`.